lib/matplotlib/colorbar.py

```diff
@@ -222,102 +224,74 @@ def _set_ticks_on_axis_warn(*args, **kw):
         """
-        This ticker needs to know the *colorbar* so that it can access
-        its *vmin* and *vmax*.  Otherwise it is the same as
-        `~.ticker.AutoLocator`.
-        """
-
-        self._colorbar = colorbar
-        nbins = 'auto'
-        steps = [1, 2, 2.5, 5, 10]
-        super().__init__(nbins=nbins, steps=steps)
-
-    def tick_values(self, vmin, vmax):
-        # flip if needed:
-        if vmin > vmax:
-            vmin, vmax = vmax, vmin
-        vmin = max(vmin, self._colorbar.norm.vmin)
-        vmax = min(vmax, self._colorbar.norm.vmax)
-        ticks = super().tick_values(vmin, vmax)
-        rtol = (vmax - vmin) * 1e-10
-        return ticks[(ticks >= vmin - rtol) & (ticks <= vmax + rtol)]
-
-
-class _ColorbarAutoMinorLocator(ticker.AutoMinorLocator):
-    """
-    AutoMinorLocator for Colorbar
-
-    This locator is just a `.AutoMinorLocator` except the min and max are
-    clipped by the norm's min and max (i.e. vmin/vmax from the
-    image/pcolor/contour object).  This is necessary so that the minorticks
-    don't extrude into the "extend regions".
-    """
-
-    def __init__(self, colorbar, n=None):
-        """
-        This ticker needs to know the *colorbar* so that it can access
-        its *vmin* and *vmax*.
+        Parameters
+        ----------
+        parent : Axes
+            Axes that specifies the position of the colorbar.
+        userax : boolean
+            True if the user passed `.Figure.colorbar` the axes manually.
         """
-        self._colorbar = colorbar
-        self.ndivs = n
-        super().__init__(n=None)
-
-    def __call__(self):
-        vmin = self._colorbar.norm.vmin
-        vmax = self._colorbar.norm.vmax
-        ticks = super().__call__()
-        rtol = (vmax - vmin) * 1e-10
-        return ticks[(ticks >= vmin - rtol) & (ticks <= vmax + rtol)]
-
-
-class _ColorbarLogLocator(ticker.LogLocator):
-    """
-    LogLocator for Colorbarbar
-
-    This locator is just a `.LogLocator` except the min and max are
-    clipped by the norm's min and max (i.e. vmin/vmax from the
-    image/pcolor/contour object).  This is necessary so ticks don't
-    extrude into the "extend regions".
 
-    """
-    def __init__(self, colorbar, *args, **kwargs):
+        if userax:
+            # copy position:
+            fig = parent.figure
+            outer_ax = fig.add_axes(parent.get_position())
+            # copy the locator if one exists:
+            outer_ax._axes_locator = parent._axes_locator
+            # if the parent is a child of another axes, swap these...
+            if (parent._axes is not None and
+                    parent in parent._axes.child_axes):
+                parent._axes.add_child_axes(outer_ax)
+                outer_ax._axes.child_axes.remove(parent)
+            else:
+                parent.remove()
+        else:
+            outer_ax = parent
+
+        inner_ax = outer_ax.inset_axes([0, 0, 1, 1])
+        self.__dict__ = inner_ax.__dict__
```

Comment:
I'm uncomfortable with this dict trick, at least in this location.  It leads to craziness like this:
```
In [64]: cbar.ax.inner_ax.inner_ax.outer_ax
Out[64]: <Axes:>
```
I think @anntzer mentioned an alternative: defining custom `__getattr__` and `__setattr__` methods to delegate to inner or outer as appropriate.

Reply:
Agree, but I don't know how to get `__getattr__` to do this more cleanly.  However, changing this to:
```python
              self.__dict__.update(inner_ax.__dict__)
```
stops the two dicts from being the same and driving the circular references you note above.